<a href="https://colab.research.google.com/github/Dharma-Ranganathan/AllAboutPython/blob/main/ML_Projects_Codes/KFold_Cross_Validation_Scores.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Cross Validation Score is to find out the proper and perfect for model selection instead train test split.

Import dependencies

In [2]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.metrics import accuracy_score

Import All classification models

In [6]:
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.ensemble import RandomForestClassifier
from sklearn.neighbors import KNeighborsClassifier

Data collection for classification model - using heart disease dataset

In [7]:
heart = pd.read_csv('/content/drive/MyDrive/Colab_python/heart_disease_data.csv')

# checking
heart.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   

   ca  thal  target  
0   0     1       1  
1   0     2       1  
2   0     2       1  
3   0     2       1  
4   0     2       1

Data Pre-processing

In [8]:

# checking if missing values exists

heart.isnull().sum()

age         0
sex         0
cp          0
trestbps    0
chol        0
fbs         0
restecg     0
thalach     0
exang       0
oldpeak     0
slope       0
ca          0
thal        0
target      0
dtype: int64

* no null values exists

In [9]:
# checking if dataset is in balanced state

heart.shape

(303, 14)

In [10]:
heart.value_counts('target')

target
1    165
0    138
Name: count, dtype: int64

* dataset is in balanced state

Basic EDA

In [12]:
# mean by target
heart.groupby('target').mean()

age       sex        cp    trestbps        chol       fbs  \
target                                                                    
0       56.601449  0.826087  0.478261  134.398551  251.086957  0.159420   
1       52.496970  0.563636  1.375758  129.303030  242.230303  0.139394   

         restecg     thalach     exang   oldpeak     slope        ca      thal  
target                                                                          
0       0.449275  139.101449  0.550725  1.585507  1.166667  1.166667  2.543478  
1       0.593939  158.466667  0.139394  0.583030  1.593939  0.363636  2.121212

In [15]:
# median by target
heart.groupby('target').median()

age  sex   cp  trestbps   chol  fbs  restecg  thalach  exang  \
target                                                                  
0       58.0  1.0  0.0     130.0  249.0  0.0      0.0    142.0    1.0   
1       52.0  1.0  2.0     130.0  234.0  0.0      1.0    161.0    0.0   

        oldpeak  slope   ca  thal  
target                             
0           1.4    1.0  1.0   3.0  
1           0.2    2.0  0.0   2.0

In [18]:
# max and min
Max = heart.groupby('target').max()
print('maximum of each feature by target :')
print(Max)
print()

Min = heart.groupby('target').min()
print('minimum of each feature by target :')
print(Min)
print()

maximum of each feature by target :
        age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
target                                                                        
0        77    1   3       200   409    1        2      195      1      6.2   
1        76    1   3       180   564    1        2      202      1      4.2   

        slope  ca  thal  
target                   
0           2   4     3  
1           2   4     3  

minimum of each feature by target :
        age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
target                                                                        
0        35    0   0       100   131    0        0       71      0      0.0   
1        29    0   0        94   126    0        0       96      0      0.0   

        slope  ca  thal  
target                   
0           0   0     0  
1           0   0     0  



Splitting features and Label

In [19]:
features = heart.drop('target',axis=1)
label = heart['target']

Train test split by first

In [23]:
x_train,x_test,y_train,y_test = train_test_split(features.values, label, test_size = 0.2, stratify=label,random_state=3)

In [24]:
print(features.shape,x_train.shape,x_test.shape)

(303, 13) (242, 13) (61, 13)


* In 303 data points, 242 is for training and 61 is for testing

Model selection by train_test_split

In [26]:
models = [LogisticRegression(max_iter=1000), SVC(kernel='linear'), KNeighborsClassifier(), RandomForestClassifier()]

In [30]:
def compare_score_train_test(models):

    for model in models:

        model.fit(x_train,y_train)

        pred = model.predict(x_test)

        accuracy = accuracy_score(pred,y_test)

        accuracy = round(accuracy,2)

        print(f'Accuracy of {model} is {accuracy}%')

In [31]:
# function call

compare_score_train_test(models)

Accuracy of LogisticRegression(max_iter=1000) is 0.79%
Accuracy of SVC(kernel='linear') is 0.77%
Accuracy of KNeighborsClassifier() is 0.66%
Accuracy of RandomForestClassifier() is 0.8%


* In train test split method way, we got to know logistic model performace and accuracy is higher compared than rest of models

Model selection by cross_val_score instead of train_test method

In [32]:
# getting new instances of models again

cross_models = [
LogisticRegression(max_iter=1000),
SVC(kernel='linear'),
KNeighborsClassifier(),
RandomForestClassifier()
]

In [40]:
# function for comparison by cv score

def compare_score_cross_validation(models):
    for model in models:
        score = cross_val_score(model,features.values,label,cv=5)

        # scores
        print(f'scores of {model} :')
        print(score)
        print()

        # mean accuracy
        accuracy = sum(score) / len(score)
        accuracy = round(accuracy,3)
        print(f'accuracy of {model} = {accuracy}%')
        print('---------------------------------------------------------------')

In [41]:
# function call
compare_score_cross_validation(cross_models)

scores of LogisticRegression(max_iter=1000) :
[0.80327869 0.86885246 0.85245902 0.86666667 0.75      ]

accuracy of LogisticRegression(max_iter=1000) = 0.828%
---------------------------------------------------------------
scores of SVC(kernel='linear') :
[0.81967213 0.8852459  0.80327869 0.86666667 0.76666667]

accuracy of SVC(kernel='linear') = 0.828%
---------------------------------------------------------------
scores of KNeighborsClassifier() :
[0.60655738 0.6557377  0.57377049 0.73333333 0.65      ]

accuracy of KNeighborsClassifier() = 0.644%
---------------------------------------------------------------
scores of RandomForestClassifier() :
[0.81967213 0.90163934 0.81967213 0.8        0.75      ]

accuracy of RandomForestClassifier() = 0.818%
---------------------------------------------------------------


* we got to know, out of 4 models, 3 models is providing same accuracy values, which means we can able to use those models in our model selection.

Note :
* Cross Validation Score is basically used only on small dataset
* Using this score , we can able to find perfect and efficient model in our training time
* More effective than train test split , but takes more time in large dataset compared to train test split method
* Using of this method could be arranged in needed only way